File: GifConvert.java
// GifConvert.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class GifConvert {

    private static boolean checkFFmpeg() {
        try {
            Process p = Runtime.getRuntime().exec(new String[]{"ffmpeg", "-version"});
            return p.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    private static List<String> buildCmd(String input, String output, int fps, String scale,
                                         double start, double duration, int quality,
                                         String filter, boolean reverse) {
        List<String> cmd = new ArrayList<>();
        cmd.add("ffmpeg");
        cmd.add("-i");
        cmd.add(input);
        cmd.add("-y");
        if (start > 0) {
            cmd.add("-ss");
            cmd.add(String.valueOf(start));
        }
        if (duration > 0) {
            cmd.add("-t");
            cmd.add(String.valueOf(duration));
        }
        List<String> filters = new ArrayList<>();
        if (fps > 0) {
            filters.add("fps=" + fps);
        }
        if (scale != null && !scale.isEmpty()) {
            filters.add("scale=" + scale);
        }
        if (filter != null && !filter.isEmpty()) {
            switch (filter) {
                case "grayscale":
                    filters.add("hue=s=0");
                    break;
                case "sepia":
                    filters.add("colorchannelmixer=.393:.769:.189:.349:.686:.168:.272:.534:.131");
                    break;
                case "negate":
                    filters.add("negate");
                    break;
            }
        }
        if (reverse) {
            filters.add("reverse");
        }
        if (!filters.isEmpty()) {
            cmd.add("-vf");
            cmd.add(String.join(",", filters));
        }
        if (quality < 1) quality = 1;
        if (quality > 31) quality = 31;
        cmd.add("-q:v");
        cmd.add(String.valueOf(quality));
        cmd.add("-c:v");
        cmd.add("gif");
        cmd.add(output);
        return cmd;
    }

    private static void convert(String input, String output, int fps, String scale,
                                double start, double duration, int quality,
                                String filter, boolean reverse) throws Exception {
        if (!checkFFmpeg()) {
            System.err.println("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.");
            System.exit(1);
        }
        List<String> cmdList = buildCmd(input, output, fps, scale, start, duration, quality, filter, reverse);
        System.out.println("Выполняется команда: " + String.join(" ", cmdList));
        ProcessBuilder pb = new ProcessBuilder(cmdList);
        pb.inheritIO();
        Process p = pb.start();
        int exitCode = p.waitFor();
        if (exitCode != 0) {
            System.err.println("Ошибка конвертации.");
            System.exit(1);
        }
        System.out.println("✅ GIF создан: " + output);
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Использование: java GifConvert <input> [--output file] [--fps N] [--scale WxH] [--start S] [--duration D] [--quality Q] [--filter F] [--reverse]");
            System.exit(1);
        }
        String input = args[0];
        String output = "output.gif";
        int fps = 10;
        String scale = null;
        double start = 0.0;
        double duration = 0.0;
        int quality = 10;
        String filter = null;
        boolean reverse = false;

        for (int i = 1; i < args.length; i++) {
            switch (args[i]) {
                case "--output":
                    if (i+1 < args.length) output = args[++i];
                    break;
                case "--fps":
                    if (i+1 < args.length) fps = Integer.parseInt(args[++i]);
                    break;
                case "--scale":
                    if (i+1 < args.length) scale = args[++i];
                    break;
                case "--start":
                    if (i+1 < args.length) start = Double.parseDouble(args[++i]);
                    break;
                case "--duration":
                    if (i+1 < args.length) duration = Double.parseDouble(args[++i]);
                    break;
                case "--quality":
                    if (i+1 < args.length) quality = Integer.parseInt(args[++i]);
                    break;
                case "--filter":
                    if (i+1 < args.length) filter = args[++i];
                    break;
                case "--reverse":
                    reverse = true;
                    break;
                default:
                    System.err.println("Неизвестный аргумент: " + args[i]);
                    System.exit(1);
            }
        }
        if (!Files.exists(Paths.get(input))) {
            System.err.println("Файл " + input + " не найден.");
            System.exit(1);
        }
        convert(input, output, fps, scale, start, duration, quality, filter, reverse);
    }
}

File: gif_convert.py
# gif_convert.py
import subprocess
import argparse
import os
import sys
import shutil
import tempfile

def check_ffmpeg():
    """Проверяет наличие ffmpeg в PATH."""
    if shutil.which("ffmpeg") is None:
        print("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.", file=sys.stderr)
        sys.exit(1)

def build_ffmpeg_cmd(input_file, output_file, fps, scale, start, duration, quality, filter_type, reverse):
    """
    Формирует команду ffmpeg на основе опций.
    Возвращает список аргументов для subprocess.
    """
    cmd = ["ffmpeg", "-i", input_file, "-y"]
    filter_parts = []

    # Обрезка по времени
    if start is not None:
        cmd.extend(["-ss", str(start)])
    if duration is not None:
        cmd.extend(["-t", str(duration)])

    # Фильтры: fps, scale, filter, reverse
    if fps is not None:
        filter_parts.append(f"fps={fps}")
    if scale is not None:
        filter_parts.append(f"scale={scale}")
    if filter_type:
        if filter_type == "grayscale":
            filter_parts.append("hue=s=0")
        elif filter_type == "sepia":
            filter_parts.append("colorchannelmixer=.393:.769:.189:.349:.686:.168:.272:.534:.131")
        elif filter_type == "negate":
            filter_parts.append("negate")
    if reverse:
        # Реверс требует сначала генерацию палитры, но мы упростим:
        # Используем фильтр reverse после всех остальных
        filter_parts.append("reverse")

    if filter_parts:
        cmd.extend(["-vf", ",".join(filter_parts)])

    # Качество GIF (q:v от 2 до 31, 2 - лучшее качество)
    if quality is not None:
        # Ограничим качество диапазоном 1-31
        q = max(1, min(31, quality))
        cmd.extend(["-q:v", str(q)])
    else:
        cmd.extend(["-q:v", "10"])

    # Кодек и выход
    cmd.extend(["-c:v", "gif", output_file])
    return cmd

def convert(input_file, output_file, fps, scale, start, duration, quality, filter_type, reverse):
    """Запускает конвертацию."""
    check_ffmpeg()
    cmd = build_ffmpeg_cmd(input_file, output_file, fps, scale, start, duration, quality, filter_type, reverse)
    print("Выполняется команда:", " ".join(cmd))
    try:
        subprocess.run(cmd, check=True)
        print(f"✅ GIF создан: {output_file}")
    except subprocess.CalledProcessError as e:
        print(f"❌ Ошибка конвертации: {e}", file=sys.stderr)
        sys.exit(1)

def main():
    parser = argparse.ArgumentParser(description="Конвертер видео в GIF с использованием FFmpeg")
    parser.add_argument("input", help="Входной видеофайл")
    parser.add_argument("--output", "-o", default="output.gif", help="Выходной GIF-файл")
    parser.add_argument("--fps", type=int, default=10, help="Частота кадров (по умолчанию 10)")
    parser.add_argument("--scale", help="Размер в формате WIDTHxHEIGHT (например, 480x320)")
    parser.add_argument("--start", type=float, help="Начало в секундах")
    parser.add_argument("--duration", type=float, help="Длительность в секундах")
    parser.add_argument("--quality", type=int, default=10, help="Качество (1-31, меньше = лучше)")
    parser.add_argument("--filter", choices=["grayscale", "sepia", "negate"], help="Применить фильтр")
    parser.add_argument("--reverse", action="store_true", help="Реверсировать анимацию")
    args = parser.parse_args()

    if not os.path.isfile(args.input):
        print(f"Файл {args.input} не найден.", file=sys.stderr)
        sys.exit(1)

    convert(args.input, args.output, args.fps, args.scale, args.start,
            args.duration, args.quality, args.filter, args.reverse)

if __name__ == "__main__":
    main()
